### PR TITLE
add ROCm devices support

### PR DESCRIPTION
**Pull request by mayank31398**
Allows the codebase to run on AMD Instinct GPUs

### CODE REVIEW

File: megatron/arguments.py [hunk 1]
```diff
@@ -641,6 +641,8 @@ def _add_training_args(parser):
                        'training if SIGTERM is received')
     group.add_argument('--tensorboard-dir', type=str, default=None,
                        help='Write TensorBoard logs to this directory.')
+    group.add_argument("--use-kernels-from-apex", action="store_true",
+                       help="use Apex kernels instead of Megatron")
```
Comment (RaymondLi0) [inline megatron/arguments.py]: In what cases is this needed?

File: megatron/fused_kernels/__init__.py [hunk 2]
```diff
@@ -1,121 +1,15 @@
...
-from torch.utils import cpp_extension
-
-# Do not override TORCH_CUDA_ARCH_LIST to allow for pre-compilation in Dockerfile
-# os.environ["TORCH_CUDA_ARCH_LIST"] = ""
+import torch
 
 
 def load(args):
+    if args.use_kernels_from_apex:
+        return
```
Comment (RaymondLi0) [inline megatron/fused_kernels/__init__.py]: I suppose this is to be implemented? Or am I missing something?

File: tools/checkpoint_loader_megatron.py [hunk 3]
```diff
@@ -30,7 +34,7 @@ def _load_checkpoint(queue, args):
         from megatron.global_vars import set_args, set_global_variables
         from megatron.checkpointing import load_args_from_checkpoint, load_checkpoint
         from megatron.model import ModelType, module
-        from megatron import mpu, fused_kernels
+        from megatron import mpu
```
Comment (RaymondLi0) [inline tools/checkpoint_loader_megatron.py]: Why is this removed?
Has it been tested that these tools still work?

**Review by RaymondLi0 — COMMENTED** [review]
Thank you for adding this Mayank!
I added a few comments, mostly to clarify.
I understand that @loubnabnl  has tested this and managed to run training on AMD GPUs?

File: megatron/arguments.py (hunk 1 above)
Comment (mayank31398) [reply megatron/arguments.py]: If a user builds apex from source, they already have the kernels to run everything.
So, its a waste to build kernels in this repo since they are just a copy of the ones in Apex.

File: megatron/fused_kernels/__init__.py (hunk 2 above)
Comment (mayank31398) [reply megatron/fused_kernels/__init__.py]: No, this is complete

File: tools/checkpoint_loader_megatron.py (hunk 3 above)
Comment (mayank31398) [reply tools/checkpoint_loader_megatron.py]: I had added it above for some reason.
reverted it.

File: megatron/arguments.py (hunk 1 above)
Comment (mayank31398) [reply megatron/arguments.py]: dropped this argument

**Comment by loubnabnl** [comment]
Yes I managed to launch the training on AMD GPUs (also tested on A100 it works fine)

**Review by RaymondLi0 — APPROVED** [review]
LGTM!
